### PR TITLE
Improve grpc_slice_buffer's grpc_slice_buffer_take_first function. 

### PR DESCRIPTION
**Pull request by sreecha**
This required modifying `grpc_slice_buffer` structure

### CODE REVIEW

File: src/core/lib/slice/slice_buffer.c [hunk 1]
```diff
@@ -156,7 +164,8 @@ void grpc_slice_buffer_pop(grpc_slice_buffer *sb) {
 
 void grpc_slice_buffer_reset_and_unref(grpc_slice_buffer *sb) {
   size_t i;
-
+  /* TODO (sreek) Ensure that it is okay not to unref slices between
```
Comment (sreecha) [inline src/core/lib/slice/slice_buffer.c]: I need to verify this some more. Not sure if this is fully safe.

File: src/core/lib/slice/slice_buffer.c [hunk 2]
```diff
@@ -147,6 +154,7 @@ void grpc_slice_buffer_addn(grpc_slice_buffer *sb, grpc_slice *s, size_t n) {
   }
 }
 
+/* TODO (sreek) - Check where this is used and if this is still safe */
 void grpc_slice_buffer_pop(grpc_slice_buffer *sb) {
```
Comment (sreecha) [inline src/core/lib/slice/slice_buffer.c]: I did not look at where all this is used and if this is safe to use.

**Comment by sreecha** [comment]
The C/C++ tests are passing on my machine. Adding 'DO NOT MERGE' label until I investigate the code (the callers) and make sure the changes I made are safe.

It would also be good to get review comments if you have time

File: src/core/lib/slice/slice_buffer.c [hunk 3]
```diff
@@ -44,29 +44,36 @@
...
-    if (sb->slices == sb->inlined) {
-      sb->slices = gpr_malloc(sb->capacity * sizeof(grpc_slice));
-      memcpy(sb->slices, sb->inlined, sb->count * sizeof(grpc_slice));
+    GPR_ASSERT(sb->capacity > slice_count);
+    if (sb->base_slices == sb->inlined) {
+      sb->base_slices = gpr_malloc(sb->capacity * sizeof(grpc_slice));
+      memcpy(sb->base_slices, sb->inlined, slice_count * sizeof(grpc_slice));
     } else {
-      sb->slices = gpr_realloc(sb->slices, sb->capacity * sizeof(grpc_slice));
+      sb->base_slices =
```
Comment (ctiller) [inline src/core/lib/slice/slice_buffer.c]: Could we do a memmove if base_slices != slices?

File: src/core/lib/slice/slice_buffer.c [hunk 4]
```diff
@@ -221,8 +243,11 @@ void grpc_slice_buffer_move_first(grpc_slice_buffer *src, size_t n,
     grpc_slice_buffer_move_into(src, dst);
     return;
   }
+
   src_idx = 0;
-  while (src_idx < src->capacity) {
+  /* This was previously written has src_idx < src->capacity. Double-check that
+   * it was a mistake */
+  while (src_idx < src->count) {
```
Comment (ctiller) [inline src/core/lib/slice/slice_buffer.c]: Suspect it was: and a good opportunity to expand tests

File: src/core/lib/slice/slice_buffer.c (hunk 3 above)
Comment (sreecha) [inline src/core/lib/slice/slice_buffer.c]: You mean doing the following as the very first line of the code?
```
if (sb->base_slices != sb->slices) {
    memmove(sb->base_slices, sb->slices, sb->count * sizeof(grpc_slice));
    sb->slices = sb->base_slices;
}
```

I was not sure what the status of the `grpc_slice`s between `base_slices` and `slices`  fields (i.e do we own them or not. Looks like we no longer own them once they are given out via the `grpc_slice_buffer_take_first` function. I will look at the code a bit more, convince myself and do it)

File: src/core/lib/slice/slice_buffer.c [hunk 5]
```diff
@@ -277,8 +302,9 @@ grpc_slice grpc_slice_buffer_take_first(grpc_slice_buffer *sb) {
   grpc_slice slice;
   GPR_ASSERT(sb->count > 0);
   slice = sb->slices[0];
-  memmove(&sb->slices[0], &sb->slices[1], (sb->count - 1) * sizeof(grpc_slice));
+  sb->slices++;
```
Comment (llhe) [inline src/core/lib/slice/slice_buffer.c]: Is this needed?
sb->slices[0] = gpr_empty_slice();

**Comment by ctiller** [comment]
What's the status of this change?

**Comment by jhseu** [comment]
Friendly ping on this change. We'd like to get this in for the TensorFlow 1.0 release, which is happening soon.

**Comment by ctiller** [comment]
Jenkins: test this please

**Pull request merged**